File: src/quest_editor/model/QuestEventModel.ts
import { QuestEventActionModel } from "./QuestEventActionModel";
import { WritableListProperty } from "../../core/observable/property/list/WritableListProperty";
import { list_property, property } from "../../core/observable";
import { ListProperty } from "../../core/observable/property/list/ListProperty";
import { Property } from "../../core/observable/property/Property";
import { WritableProperty } from "../../core/observable/property/WritableProperty";
import { defined, require_integer, require_non_negative_integer } from "../../core/util";
import { WaveModel } from "./WaveModel";

export class QuestEventModel {
    private readonly _delay: WritableProperty<number>;
    private readonly _actions: WritableListProperty<QuestEventActionModel> = list_property();

    readonly id: number;
    readonly section_id: number;
    readonly wave: WaveModel;
    readonly delay: Property<number>;
    readonly actions: ListProperty<QuestEventActionModel> = this._actions;
    readonly unknown: number;

    constructor(id: number, section_id: number, wave: WaveModel, delay: number, unknown: number) {
        require_non_negative_integer(id, "id");
        require_non_negative_integer(section_id, "section_id");
        defined(wave, "wave");
        require_non_negative_integer(delay, "delay");
        require_integer(unknown, "unknown");

        this.id = id;
        this.section_id = section_id;
        this.wave = wave;
        this._delay = property(delay);
        this.delay = this._delay;
        this.unknown = unknown;
    }

    set_delay(delay: number): void {
        require_non_negative_integer(delay, "delay");
        this._delay.val = delay;
    }

    add_action(action: QuestEventActionModel): void {
        this._actions.push(action);
    }
}

File: web/src/main/kotlin/world/phantasmal/web/questEditor/models/QuestEventModel.kt
package world.phantasmal.web.questEditor.models

import world.phantasmal.cell.Cell
import world.phantasmal.cell.list.ListCell
import world.phantasmal.cell.list.SimpleListCell
import world.phantasmal.cell.map
import world.phantasmal.cell.mutableCell

class QuestEventModel(
    id: Int,
    val areaId: Int,
    sectionId: Int,
    waveId: Int,
    delay: Int,
    val unknown: Int,
    actions: MutableList<QuestEventActionModel>,
) {
    private val _id = mutableCell(id)
    private val _sectionId = mutableCell(sectionId)
    private val _waveId = mutableCell(waveId)
    private val _delay = mutableCell(delay)
    private val _actions = SimpleListCell(actions)

    val id: Cell<Int> = _id
    val sectionId: Cell<Int> = _sectionId
    val wave: Cell<WaveModel> = map(_waveId, _sectionId) { id, sectionId ->
        WaveModel(id, areaId, sectionId)
    }
    val delay: Cell<Int> = _delay
    val actions: ListCell<QuestEventActionModel> = _actions

    fun setId(id: Int) {
        _id.value = id
    }

    fun setSectionId(sectionId: Int) {
        _sectionId.value = sectionId
    }

    fun setWaveId(waveId: Int) {
        _waveId.value = waveId
    }

    fun setDelay(delay: Int) {
        _delay.value = delay
    }

    fun addAction(action: QuestEventActionModel) {
        _actions.add(action)
    }

    fun addAction(index: Int, action: QuestEventActionModel) {
        _actions.add(index, action)
    }

    fun removeAction(action: QuestEventActionModel) {
        _actions.remove(action)
    }
}
